readium/shared/src/main/java/org/readium/r2/shared/util/downloads/foreground/ForegroundDownloadManager.kt

```diff
@@ -0,0 +1,136 @@
+            .orEmpty()
+
+        val listenersForId = listeners[id].orEmpty()
+
+        when (response) {
+            is Try.Success -> {
+                withContext(Dispatchers.IO) {
+                    try {
+                        val dest = File.createTempFile(
+                            UUID.randomUUID().toString(),
+                            dottedExtension
+                        )
+                        dest.writeBytes(response.value.body)
+                    } catch (e: Exception) {
+                        val error = DownloadManager.Error.FileError(ThrowableError(e))
+                        listenersForId.forEach { it.onDownloadFailed(id, error) }
+                    }
+                }
+            }
+            is Try.Failure -> {
+                val error = mapError(response.value)
+                listenersForId.forEach { it.onDownloadFailed(id, error) }
+            }
+        }
+
+        listeners.remove(id)
+    }
+
+    private fun mapError(httpException: HttpException): DownloadManager.Error {
+        val httpError = ThrowableError(httpException)
+        return when (httpException.kind) {
+            HttpException.Kind.MalformedRequest ->
+                DownloadManager.Error.Unknown(httpError)
+            HttpException.Kind.MalformedResponse ->
+                DownloadManager.Error.HttpData(httpError)
+            HttpException.Kind.Timeout ->
+                DownloadManager.Error.Unreachable(httpError)
+            HttpException.Kind.BadRequest ->
+                DownloadManager.Error.Unknown(httpError)
+            HttpException.Kind.Unauthorized ->
+                DownloadManager.Error.Forbidden(httpError)
+            HttpException.Kind.Forbidden ->
+                DownloadManager.Error.Forbidden(httpError)
+            HttpException.Kind.NotFound ->
+                DownloadManager.Error.NotFound(httpError)
+            HttpException.Kind.ClientError ->
+                DownloadManager.Error.HttpData(httpError)
+            HttpException.Kind.ServerError ->
+                DownloadManager.Error.Server(httpError)
+            HttpException.Kind.Offline ->
+                DownloadManager.Error.Unreachable(httpError)
+            HttpException.Kind.Cancelled ->
+                DownloadManager.Error.Unknown(httpError)
+            HttpException.Kind.Other ->
+                DownloadManager.Error.Unknown(httpError)
+        }
+    }
+
+    public override fun cancel(requestId: DownloadManager.RequestId) {
+        jobs.remove(requestId)?.cancel()
+        val listenersForId = listeners[requestId].orEmpty()
+        listenersForId.forEach { it.onDownloadCancelled(requestId) }
+        listeners.remove(requestId)
+    }
+
+    public override fun register(
+        requestId: DownloadManager.RequestId,
+        listener: DownloadManager.Listener
+    ) {
+        listeners.getOrPut(requestId) { mutableListOf() }.add(listener)
+    }
+
+    public override fun close() {
```

Comment:
Maybe we can cancel all the known requests in `close`, to let the opportunity to recover gracefully in the listeners?